doc/api-management.md

```diff
@@ -0,0 +1,59 @@
+# API Management
+
+> Guidance below for Theia framework. End products will get better development experience by using private visibility.
+
+Theia is a framework embracing openness, extensibility, and customizability as much as possible:
+- API defaults to public for clients.
+- API defaults to protected for extenders.
+- Language constructions prohibiting runtime access to internals are never used and so on.
+
+Usually, the version management is built around API visibility.
+Particularly, if a public API is broken then the major release is required.
+
+Following the conventional approach is not practicable for Theia since all APIs are more or less public.
+It will slow down API innovation, accumulate technical debt or require many major releases.
+
+Instead, we require the major release only if a stable API is broken.
+Breaking an experimental API can be done with the minor release,
+but not necessary if such API does not have sufficient adoption.
+
+## Stability
+
+Conceptually, an API is all assumptions adopters depend on to make use or extend Theia.
+A stable API is based on assumptions that are not subject to change.
+It does not matter whether an API is public or internal, or whether it is used a lot or never.
+
+> For instance, the language server protocol depends on such data types like URIs and positions
+since they don't change from language to language. It ensures its stability.
+
+The API stability indicated by adding, exclusively, `@experimental` or `@stable` js-doc tags.
+An API without a stability tag considered to be experimental.
+
+### Experimental
+
+- All new APIs should always be added as `experimental`
+since it's almost impossible to get stable API right first time.
+- Usually, experimental APIs don't require the stability tag,
+but if a new member is added to `stable` API then it should be explicitly annotated.
+- Such APIs should not have extensive documentation to avoid the maintenance burden.
```

Comment:
`Such APIs should not have extensive documentation`

I personally do not agree with this. Can we leave this out?

Imagine the following, someone introduces a new API, which is experimental, so it is not documented extensively. Then after some time, the API becomes stable; will you or the author of the API go back and do nothing but write documentation? It sounds good but it is far from reality.

Reply:
My point is that noone should expect that docs are up-to-date, asks during reviews and so on. If they want to have such level they should ask to make such API stable, then we will care about proper docs and backward compatibility as per semver. If some concrete developer wants to put effort initially it is up to him. I will try to rephrase it.

Reply:
That's true. Avoiding maintenance burden should not be an excuse for not writing documentation. IMHO good documentation is necessary if we want to see this framework more widely adopted.

We could approach it like this.

 * For experimental API, the amount of documentation should be well balanced: document the parts that are not obvious, but be careful where changes might be expected in the future.
 * Graduation of an API to stable involves a review of its documentation.

Reply:
> Then after some time, the API becomes stable; will you or the author of the API go back and do nothing but write documentation? It sounds good but it is far from reality.

That's a requirement for stable API. There is a GH  issue to make some API stable. We take time to look at API again, make sure that it is not going to be changed next year and document it. If it was not done API does not become stable.